tests/tx_validation_test.go

```diff
@@ -361,6 +361,172 @@ func TestValidationPoolInsertMagma(t *testing.T) {
+
+	var txs types.Transactions
+	// deploy a contract for contract execution tx type
+	{
+		values := map[types.TxValueKeyType]interface{}{
+			types.TxValueKeyNonce:         reservoir.GetNonce(),
+			types.TxValueKeyFrom:          reservoir.GetAddr(),
+			types.TxValueKeyTo:            (*common.Address)(nil),
+			types.TxValueKeyAmount:        big.NewInt(0),
+			types.TxValueKeyGasLimit:      gasLimit,
+			types.TxValueKeyGasPrice:      big.NewInt(25 * params.Gkei),
+			types.TxValueKeyHumanReadable: false,
+			types.TxValueKeyData:          common.FromHex(code),
+			types.TxValueKeyCodeFormat:    params.CodeFormatEVM,
+		}
+
+		tx, err := types.NewTransactionWithMap(types.TxTypeSmartContractDeploy, values)
+		assert.Equal(t, nil, err)
+
+		err = tx.SignWithKeys(signer, reservoir.Keys)
+		assert.Equal(t, nil, err)
+
+		txs = append(txs, tx)
+
+		contract.Addr = crypto.CreateAddress(reservoir.Addr, reservoir.Nonce)
+
+		reservoir.AddNonce()
+	}
+
+	// set code for contract execution tx type
+	{
+		auth, err := types.SignAuth(&types.Authorization{
+			ChainID: bcdata.bc.Config().ChainID,
+			Address: contract.Addr,
+			Nonce:   uint64(0),
+		}, eoaWithCode.Keys[0])
+		assert.Equal(t, nil, err)
+
+		authorizationList := types.AuthorizationList{*auth}
+
+		tx := types.NewMessage(reservoir.Addr, &eoaWithCode.Addr, reservoir.GetNonce(), nil, gasLimit,
+			nil, big.NewInt(25*params.Gkei), big.NewInt(25*params.Gkei), nil, false, uint64(0), nil, authorizationList)
+		tx.ChainId().Set(bcdata.bc.Config().ChainID)
+		err = tx.SignWithKeys(signer, reservoir.Keys)
+		assert.Equal(t, nil, err)
+
+		txs = append(txs, tx)
+
+		reservoir.AddNonce()
+		eoaWithCode.AddNonce()
+		accountMap.IncNonce(eoaWithCode.Addr)
+	}
+
+	if err := bcdata.GenABlockWithTransactions(accountMap, txs, prof); err != nil {
+		t.Fatal(err)
+	}
+
+	// make TxPool to test validation in 'TxPool add' process
+	txpool := blockchain.NewTxPool(blockchain.DefaultTxPoolConfig, bcdata.bc.Config(), bcdata.bc)
+
+	// test for all tx types
+	for _, testTxType := range testTxTypes {
+		txType := testTxType.txType
+
+		// generate invalid txs and check the return error
+		for _, invalidCase := range invalidCases {
+			to := reservoir
+			if toBasicType(testTxType.txType) == types.TxTypeSmartContractExecution {
+				to = contract
+			}
+
+			// generate a new tx and mutate it
+			valueMap, _ := genMapForTxTypes(reservoir, to, txType)
+			invalidMap, expectedErr := invalidCase.fn(txType, valueMap, eoaWithCode.Addr)
+
+			tx, err := types.NewTransactionWithMap(txType, invalidMap)
+			assert.Equal(t, nil, err)
+
+			if from, _ := tx.From(); from == eoaWithCode.Addr {
+				err = tx.SignWithKeys(signer, eoaWithCode.Keys)
+				assert.Equal(t, nil, err)
+			} else {
+				err = tx.SignWithKeys(signer, reservoir.Keys)
+				assert.Equal(t, nil, err)
+			}
+
+			if txType.IsFeeDelegatedTransaction() {
+				tx.SignFeePayerWithKeys(signer, reservoir.Keys)
+				assert.Equal(t, nil, err)
+			}
```

Comment:
We may concise this singing error check by refactoring the if else condition and results in one `assert`.

Reply:
The above conditions are contradictory, so I agree with you. Thank you.
On the other hand, this part is a different condition, so it would be better to assert it separately.
```
			if txType.IsFeeDelegatedTransaction() {
				err = tx.SignFeePayerWithKeys(signer, reservoir.Keys)
				assert.Equal(t, nil, err)
			}
```

Reply:
It's fixed by https://github.com/kaiachain/kaia/pull/145/commits/f2cded5a6f68af069598656747a163939bd14b38.
Please check it!